balancer_conn_wrappers.go

```diff
@@ -20,136 +20,172 @@ package grpc
-					}
-				}
+			update := u.(watcherUpdate)
+			switch update.typ {
+			case updateTypeClientConnState:
+				ccb.handleClientConnStateChange(update.update.(*balancer.ClientConnState))
+			case updateTypeSubConnState:
+				ccb.handleSubConnStateChange(update.update.(*scStateUpdate))
+			case updateTypeExitIdle:
+				ccb.handleExitIdle()
+			case updateTypeResolverError:
+				ccb.handleResolverError(update.update.(error))
+			case updateTypeSwitchTo:
+				ccb.handleSwitchTo(update.update.(string))
+			case updateTypeSubConn:
+				ccb.handleRemoveSubConn(update.update.(*acBalancerWrapper))
 			default:
-				logger.Errorf("ccBalancerWrapper.watcher: unknown update %+v, type %T", t, t)
+				logger.Errorf("ccBalancerWrapper.watcher: unknown update %+v, type %T", update, update)
 			}
 		case <-ccb.closed.Done():
 		}
 
 		if ccb.closed.HasFired() {
-			ccb.balancerMu.Lock()
-			ccb.balancer.Close()
-			ccb.balancerMu.Unlock()
-			ccb.mu.Lock()
-			scs := ccb.subConns
-			ccb.subConns = nil
-			ccb.mu.Unlock()
-			ccb.UpdateState(balancer.State{ConnectivityState: connectivity.Connecting, Picker: nil})
-			ccb.done.Fire()
-			// Fire done before removing the addr conns.  We can safely unblock
-			// ccb.close and allow the removeAddrConns to happen
-			// asynchronously.
-			for acbw := range scs {
-				ccb.cc.removeAddrConn(acbw.getAddrConn(), errConnDrain)
-			}
+			ccb.handleClose()
 			return
 		}
 	}
 }
 
-func (ccb *ccBalancerWrapper) close() {
-	if ccb == nil {
-		return
+// updateClientConnState is invoked by grpc to push a ClientConnState update to
+// the underlying balancer.
+//
+// Unlike other methods invoked by grpc to push updates to the underlying
+// balancer, this method cannot simply push the update onto the update channel
+// and return. It needs to return the error returned by the underlying balancer
+// back to grpc which propagates that to the resolver.
+func (ccb *ccBalancerWrapper) updateClientConnState(ccs *balancer.ClientConnState) error {
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeClientConnState, update: ccs})
+	res := <-ccb.resultCh.Get()
+	ccb.resultCh.Load()
+	// If the returned error is nil, attempting to type assert to error leads to
+	// panic. So, this needs to handled separately.
+	if res == nil {
+		return nil
 	}
-	ccb.closed.Fire()
-	<-ccb.done.Done()
+	return res.(error)
 }
 
-func (ccb *ccBalancerWrapper) exitIdle() bool {
-	if ccb == nil {
-		return true
-	}
-	if !ccb.hasExitIdle {
-		return false
+// handleClientConnStateChange handles a ClientConnState update from the update
+// channel and invokes the appropriate method on the underlying balancer.
+//
+// If the addresses specified in the update contain addresses of type "grpclb"
+// and the selected LB policy is not "grpclb", these addresses will be filtered
+// out and ccs will be modified with the updated address list.
+func (ccb *ccBalancerWrapper) handleClientConnStateChange(ccs *balancer.ClientConnState) {
+	ccb.balancerMu.Lock()
```

Comment:
Why does this hold balancerMu?

Reply:
Because it calls `ccb.balancer.UpdateClientConnState` in the last line. Would you rather that I grab the lock around that statement?